### PR TITLE
Centralise IP address object address retrieval and existence checks.

### PR DESCRIPTION
**Pull request by citrus-it**
Following [illumos 16677](https://www.illumos.org/issues/16677), the error message that indicates that
an address object does not exist has been changed to be
consistent, whereas it was previously one of two different messages
or sometimes even completely blank.

A couple of places in omicron relied on the specific error
message so this change centralises the check, and updates it
to cater for illumos pre- and post-16677. Longer term this
could be replaced by bindings to libipadm.

### CODE REVIEW

File: illumos-utils/src/ipadm.rs [hunk 1]
```diff
@@ -71,77 +103,53 @@ impl Ipadm {
...
-                    "static",
-                    "-a",
-                    &listen_addr.to_string(),
-                    &addrobj,
-                ]);
-                execute(cmd).map(|_| ())
-            }
-            Err(other) => Err(other),
-            Ok(_) => Ok(()),
+        if !Self::addrobj_exists(&addrobj)? {
```
Comment (bnaecker) [inline illumos-utils/src/ipadm.rs:124]: Same note here.

File: illumos-utils/src/ipadm.rs [hunk 2]
```diff
@@ -71,77 +103,53 @@ impl Ipadm {
...
-                    "-t",
-                    "-T",
-                    "dhcp",
-                    &addrobj,
-                ]);
-                execute(cmd)?;
-            }
-            Ok(_) => (),
-        };
+        if !Self::addrobj_exists(&addrobj)? {
```
Comment (bnaecker) [inline illumos-utils/src/ipadm.rs:147]: Same note here.

**Review by karencfv — COMMENTED** [review]
Thanks for the clean up and updating the error messages :) 

Do you think it's possible in any way to rely on exit status instead of error messages? If not, that's totally fine. It  just makes me worried that we'll miss a change like this at some point 🤔

File: illumos-utils/src/ipadm.rs [hunk 3]
```diff
@@ -37,6 +41,34 @@ impl Ipadm {
         }
     }
 
+    pub fn addrobj_addr(
+        addrobj: &str,
+    ) -> Result<Option<String>, ExecutionError> {
```
[the remark this answers is not in the document]
Comment (citrus-it) [reply illumos-utils/src/ipadm.rs:46]: Yes, let's do that instead.

**Comment by citrus-it** [comment]
> Do you think it's possible in any way to rely on exit status instead of error messages? If not, that's totally fine. It just makes me worried that we'll miss a change like this at some point 🤔

We should probably refactor things here to call into libipadm then we can [directly check the response codes](https://github.com/oxidecomputer/illumos-gate/blob/stlouis/usr/src/lib/libipadm/common/libipadm.h#L71) instead of attempting to parse stderr from `ipadm`. However, libipadm is currently an unstable/uncomitted interface (that is, can change at any time) but then so is the error output of `ipadm`. We could perhaps request that the libipadm interfaces we need be stablised in upstream illumos.

**Review by karencfv — APPROVED** [review]
Looks good, thanks!

**Comment by karencfv** [comment]
> We should probably refactor things here to call into libipadm then we can [directly check the response codes](https://github.com/oxidecomputer/illumos-gate/blob/stlouis/usr/src/lib/libipadm/common/libipadm.h#L71) instead of attempting to parse stderr from ipadm. However, libipadm is currently an unstable/uncomitted interface (that is, can change at any time) but then so is the error output of ipadm. We could perhaps request that the libipadm interfaces we need be stablised in upstream illumos.

That would be really good!

**Comment by citrus-it** [comment]
I've tested this by bringing up the control plane on a bench gimlet, confirming that restarting the `zone-setup` service is successful without changing existing interfaces, and that `zone-bundle` works, selecting the correct address and falling back to localhost if the addrobj does not exist, and the expected error if I manually change it to be an IPv4 address:

```
gimlet-sn06 # ipadm delete-addr underlay0/sled6
gimlet-sn06 # ipadm create-addr -T static -a local=192.168.2.1/24 underlay0/sled6
gimlet-sn06 # /tmp/zone-bundle list
Error: failed to fetch underlay address

Caused by:
    Unexpectedly got IPv4 address for underlay0/sled6: 192.168.2.1
```

**Review by bnaecker — APPROVED** [review]
Looks great, thanks!

**Pull request merged**